server_structs/director.go

```diff
@@ -141,8 +148,131 @@ type (
+		CollectionsUrl *url.URL
+	}
+
+	XPelTokGen struct {
+		Issuers       []*url.URL
+		MaxScopeDepth uint
+		Strategy      StrategyType
+		BasePaths     []string
+		VaultServer   *url.URL
+	}
+
+	DirectorResponse struct {
+		ObjectServers []*url.URL // List of servers provided in Link header
+		Location      *url.URL   // URL content of the location header
+		XPelAuthHdr   XPelAuth
+		XPelNsHdr     XPelNs
+		XPelTokGenHdr XPelTokGen
+	}
 )
 
+func (x XPelNs) GetName() string {
+	return "X-Pelican-Namespace"
+}
+func (x *XPelNs) ParseRawResponse(resp *http.Response) error {
+	raw := resp.Header.Values(x.GetName())
+	if len(raw) == 0 {
+		return errors.Errorf("No %s header found.", x.GetName())
+	}
+	keyDict := utils.HeaderParser(raw[0])
+	x.Namespace = keyDict["namespace"]
+	x.RequireToken, _ = strconv.ParseBool(keyDict["require-token"])
+	if keyDict["collections-url"] != "" {
+		x.CollectionsUrl, _ = url.Parse(keyDict["collections-url"])
+	}
+	return nil
+}
+
+func (x XPelAuth) GetName() string {
+	return "X-Pelican-Authorization"
+}
+func (x *XPelAuth) ParseRawResponse(resp *http.Response) error {
+	// If the director provides an auth header, raw will have an array of length 1.
+	raw := resp.Header.Values(x.GetName())
+	if len(raw) > 0 {
+		x.Issuers = make([]*url.URL, 0)
+		// clean up the string and split it by commas to fetch each issuer. Can't use
+		// utils.HeaderParser, because we don't have unique keys here.
+		cleaned := strings.ReplaceAll(raw[0], " ", "")
+		issuers := strings.Split(cleaned, ",")
+		for _, issuer := range issuers {
+			issuerUrlStr := strings.TrimPrefix(issuer, "issuer=")
+			issuerUrl, err := url.Parse(issuerUrlStr)
+			if err != nil {
+				return errors.Errorf("Failed to parse issuer URL %s from Director's %s header: %v", issuerUrlStr, x.GetName(), err)
+			}
+			x.Issuers = append(x.Issuers, issuerUrl)
+		}
+	}
+	return nil
+}
+
+func (x XPelTokGen) GetName() string {
+	return "X-Pelican-Token-Generation"
+}
+func (x *XPelTokGen) ParseRawResponse(resp *http.Response) error {
+	raw := resp.Header.Values(x.GetName())
+	if len(raw) > 0 {
+		// Parse issuer, for now assuming a single value but eventually may be multiple
+		x.Issuers = make([]*url.URL, 0)
+		keyDict := utils.HeaderParser(raw[0])
+		issuerUrl, err := url.Parse(keyDict["issuer"])
+		if err != nil {
+			return errors.Errorf("Failed to parse issuer URL %s from Director's %s header: %v", keyDict["issuer"], x.GetName(), err)
+		}
+		x.Issuers = append(x.Issuers, issuerUrl)
+
+		// Parse scope depth
+		maxScopeDepth, err := strconv.ParseUint(keyDict["max-scope-depth"], 10, 32)
+		if err != nil {
+			return errors.Errorf("Failed to parse max-scope-depth %s from Director's %s header: %v", keyDict["max-scope-depth"], x.GetName(), err)
+		}
+		x.MaxScopeDepth = uint(maxScopeDepth)
+
+		// Parse strategy
+		strategy, exists := keyDict["strategy"]
+		if !exists {
+			return errors.Errorf("No credential generation strategy found in Director's %s header", x.GetName())
```

Comment:
Do we always have a strategy? I'm unsure if this is an error or not.

Reply:
If we don't have a strategy, the client can't know how to generate the token (Does it visit a well-known URI? Does it drive to the issuer's house?) so it _should_ be an error. If the Director side doesn't acknowledge that by responding with some error code and logging something, let's open it as another ticket.